### PR TITLE
Hybrid Pseudohamiltonian Implementation using JaQMC

### PR DESCRIPTION
**Pull request by revpba**
## Summary

Implemented Hybrid Pseudopotential Hamiltonian (HPH) support in JaQMC based on the method proposed by Krogel & Reboredo (2020):

> Krogel, J. T., & Reboredo, F. A. (2020). Hybridizing pseudo-Hamiltonians and non-local pseudopotentials in diffusion Monte Carlo. J. Chem. Phys. 153, 104105.

The implementation follows the hybridization idea by taking the residual nonlocal channels (s and p orbitals) computed in the existing ECP workflow in JaQMC, and combining them with the local channel and L2 channel computed in the existing PH workflow. In this way, the HPH formulation is constructed directly from the current ECP and PH components already available in JaQMC.

## Motivation

The current pseudopotential implementation in JaQMC supports ECP and PH formalisms separately, but does not include an implementation of the Hybrid Pseudopotential Hamiltonian (HPH).

This update enables HPH calculations by combining the existing ECP and PH workflows, allowing hybrid pseudopotential evaluations whenever the required pseudopotential data are available (e.g., NWChem-format files for ECP and XML files for PH).

This extension makes it possible to perform calculations for systems where hybrid pseudopotentials are required, such as transition-metal systems including Sc ionization energy and ScO dissociation energy.

## Implementation

- [ ] XML pseudopotential parsing

Added functionality to read pseudopotential XML files and extract the local and L2 channels into arrays containing semilocal radial potential values. This provides a more straightforward representation of the PH data compared with the previous implementation.

- [ ] Hybrid Hamiltonian construction

Implemented the HPH formulation by combining:

- the local and L2 channels computed within the existing PH workflow, and
- the residual nonlocal channels (e.g. s and p) computed within the existing ECP workflow.

- [ ] Configuration updates

Added new configuration files for transition-metal systems:

- X.py for transition-metal atoms
- XO.py for transition-metal oxides

Separate configuration files were also introduced specifically for HPH calculations.

- [ ] Example pseudopotentials

Added example pseudopotential data for Sc, including:

- ccECP pseudopotentials from pseudopotential.org
- truncated forms of the hybrid pseudo-Hamiltonian (n = 2, …, 7)
- XML files containing the local (d), L2, s, and p channels required for the HPH construction.

- [ ] Execution workflow improvements

- Revised run.py so calculations can resume automatically from the latest checkpoint when available, while still starting a new run if no checkpoint is found.
- Adjusted the VMC workflow accordingly to support this resume logic.
- Added a Makefile to simplify running example calculations.

- [ ] Supporting utilities

- Extended special.py with additional Legendre polynomials required for truncated HPH calculations.

## Testing

The implementation was tested using **Sc first ionization energy** and **ScO dissociation energy** calculations with multiple pseudopotential types, including:

- ccECP [https://pseudopotentiallibrary.org/recipes/Sc/ccECP/Sc.ccECP.nwchem](url)
- **Hybrid pseudopotential (HPH)**
- **L2**
- **Truncated HPH potentials (orders n = 2–7)**

### Reproducibility

The tests can be reproduced with the following workflow.

1. Prepare a conda environment with the required libraries: JaQMC, LapNet, LapJAX, kfac-jax, JAX and other related dependencies
2. Prepare configuration files for each system.

For **Sc ionization energy**:
- run calculations for Sc and Sc⁺
- specify the spin configuration, charge, and pseudopotential

For **ScO dissociation energy**:
- run calculations for O (all-electron), Sc, and ScO
- specify the spin configuration, Sc–O bond distance, and pseudopotential

3. Simulation launched using the provided makefile
4. Final energies are obtained using reblocking analysis [https://pyblock.readthedocs.io/en/latest/](url) and convergence criteria for optimization

Base config for LapNet is provided in:
[base_config.py](https://github.com/user-attachments/files/25847341/base_config.py)

### Results

The workflow successfully runs HPH-based calculations and produces stable energies across different pseudopotential types.

For the **Sc ionization energy**, the HPH result shows good agreement with the **NIST reference value**, with a deviation of approximately **9.95 meV**.

For **ScO dissociation energy**, the HPH result shows a deviation of approximately **0.41 eV** from the **RCCSD(T) reference value** computed in Molpro.

Graphical comparisons are provided in:

- `Sc-ionization.png`
- `ScO-dissociation.png`

Sc-ionization
ScO-dissociation

## Additional Notes

The implementation and testing were performed at the **Hongo Laboratory, Japan Advanced Institute of Science and Technology (JAIST)**

Feedback or discussion regarding the **ScO dissociation energy results**, especially in comparison with the reference calculations, would be greatly appreciated.

### CODE REVIEW

File: makefile [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# Makefile for O
```
Comment (WeiluoRen-ByteDance) [inline makefile:1]: why this file is on the top  level? maybe move to examples instead?

File: run.py [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+
+'''
+A show case for LapNet with PH and ECP.
+
+LapNet: https://github.com/bytedance/LapNet
+'''
+
+import os
```
Comment (WeiluoRen-ByteDance) [inline run.py:13]: how is it different from the new run.py file in examples/pp/lapnet?

File: jaqmc/pp/hamiltonian.py [hunk 3]
```diff
@@ -415,12 +415,17 @@ def pp_energy(f: WavefunctionLike,
       ph_atoms = set(pp_cfg.ph_info[1].keys())
   logging.info(f'Elements for Pseudo-Hamiltonian: {ph_atoms}')
 
+  ph_mode = getattr(pp_cfg, 'ph_mode', None) # get ph_mode from pp_cfg
+  use_ecp_for_ph_atoms = (ph_mode == "hybrid") # if ph_mode is hybrid, use ECP for PH
```
Comment (WeiluoRen-ByteDance) [inline jaqmc/pp/hamiltonian.py:419]: what do we mean by "use ECP for PH"?

Comment (WeiluoRen-ByteDance) [reply jaqmc/pp/hamiltonian.py:419]: you mean for elements configured to use PH, we actually treat it with ECP calculations?

File: jaqmc/loss/vmc.py [hunk 4]
```diff
@@ -172,7 +172,15 @@ def total_energy_jvp(primals, tangents):
...
     primals_out = loss, (func_state, aux_data)
 
-    tangents_out = (jnp.dot(psi_tangent, diff) / device_batch_size, (func_state, aux_data))
+    # Ensure that the tangent returned by the JVP computation matches the primal loss in both shape and dtype
+    # Since the VMC loss is scalar, we explicitly reshape and 
+    # cast the tangent to avoid potential mismatches introduced by JAX transformations
+    loss_tangent = jnp.dot(psi_tangent, diff) / device_batch_size
+    loss = jnp.asarray(loss)
+    loss_tangent = jnp.asarray(loss_tangent)
+    loss_tangent = jnp.reshape(loss_tangent, loss.shape).astype(loss.dtype)
```
Comment (WeiluoRen-ByteDance) [inline jaqmc/loss/vmc.py:181]: So you mention that "Ensure that the tangent returned by the JVP computation matches the primal loss in both shape and dtype". Just curious under which circumstances, would it not be the case?

Comment (WeiluoRen-ByteDance) [reply jaqmc/loss/vmc.py:181]: In other words, if we don't make this change, then in which scenario will there be a "tanget not consistent with primal" issue?

File: makefile (hunk 1 above)
Comment (revpba) [reply makefile:1]: The Makefile was intended to make it easier for users to run JaQMC. However, I agree that placing it at the top level may not be the best choice.

File: run.py (hunk 2 above)
Comment (revpba) [reply run.py:13]: It is actually the same run.py file. My intention was to keep it identical to the original JaQMC version, but I placed it in the root directory for convenience. 

However, I realize this may not be the appropriate location, and it would be better to keep it in its original place. I apologize for this oversight.

File: jaqmc/pp/hamiltonian.py (hunk 3 above)
Comment (revpba) [reply jaqmc/pp/hamiltonian.py:419]: I apologize for the confusion earlier. What I intended to explain is how the hybrid pseudopotential (HPH) is constructed in the implementation.

According to the equation in the paper below: 

HPH

The HPH potential consists of three components: the local term v_loc, the L_2 term v_L2, and the nonlocal channels v_l. In our implementation, the local and L_2 parts are evaluated using the Pseudo-Hamiltonian (PH) framework, while the remaining nonlocal channels are evaluated using the ECP framework.

Therefore, the total hybrid potential is obtained by combining the PH contribution with the residual nonlocal contribution v_l computed via ECP. To enable this, the element must also be included in the ecp_element_list, so that the nonlocal part can be evaluated by the ECP

`  for sym, coord in pyscf_mol._atom:`
`      if sym in pyscf_mol._ecp:`
`        if sym not in ph_atoms or use_ecp_for_ph_atoms:`
`          ecp_atoms.append((sym, coord))`
`          ecp_element_list.append(sym)`

File: jaqmc/loss/vmc.py (hunk 4 above)
Comment (revpba) [reply jaqmc/loss/vmc.py:181]: I apologize that I don't recall the exact situation but I was resuming the calculation based on the last checkpoint file I made. The current implementation assumes that the 

`jnp.dot(psi_tangent, diff) / device_batch_size`

returns scalar with the same shape and dtype as loss. I previously encountered a case in which those are different. So it is likely to keep the original implementation if this assumptions always holds.

**Comment by WeiluoRen-ByteDance** [comment]
@revpba Hey Reva, please remove the top level makefile and run.py files. I appreciate that having such makefile would be convenient to run some examples. Maybe we can have a separate PR to address it more properly .

File: jaqmc/loss/vmc.py (hunk 4 above)
Comment (WeiluoRen-ByteDance) [reply jaqmc/loss/vmc.py:181]: @Byte-RuichenLi please take a look

File: examples/pp/lapnet/configs/ecp/X.py [hunk 5]
```diff
@@ -10,18 +10,20 @@
...
     mol = gto.Mole()
-    # Set up molecule
     mol.build(
-        atom=f'{symbol} 0 0 0',
-        basis={symbol: 'ccecpccpvdz'},
-        ecp={symbol: 'ccecp'},
-        spin=spin)
+        atom=f"{symbol} 0 0 0",
+        basis={symbol: "ccpvdz"},
+        spin=spin,
```
Comment (WeiluoRen-ByteDance) [inline examples/pp/lapnet/configs/ecp/X.py:22]: hmm does your change here break the vanilla ECP code path?

File: examples/pp/lapnet/configs/hph/XO.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    for vps in vps_list:
+        if vps.get("l") == l_target:
+            data_tag = vps.find(".//radfunc//data")
+            if data_tag is None:
+                raise RuntimeError(f"No radial function data found for vps(l={l_target}).")
+
+            text = data_tag.text
+            return parse_compact_data(text)
+
+    raise RuntimeError(f"No semilocal vps entry found for angular momentum l={l_target}.")
```
Comment (WeiluoRen-ByteDance) [inline examples/pp/lapnet/configs/hph/XO.py:169]: Can you double check if we can move the logic here to a new file in directory jaqmc/pp? Take a look at this file https://github.com/bytedance/jaqmc/blob/main/jaqmc/pp/ph/data.py. I feel they are analogous.

Comment (WeiluoRen-ByteDance) [reply examples/pp/lapnet/configs/hph/XO.py:169]: Right now we are repeating this logic in every single config files, which does not look good

File: examples/pp/lapnet/configs/ecp/X.py (hunk 5 above)
Comment (revpba) [reply examples/pp/lapnet/configs/ecp/X.py:22]: I apologize for overlooking the update to examples/pp/lapnet/configs/ecp/X.py. I will fix this issue.

File: examples/pp/lapnet/configs/hph/XO.py (hunk 6 above)
Comment (revpba) [reply examples/pp/lapnet/configs/hph/XO.py:169]: Alright, Weiluo. I will check jaqmc/pp/ph/data.py. I also realize that there are already functions for parsing XML data similar to those used in my current config files.

**Comment by revpba** [comment]
> @revpba Hey Reva, please remove the top level makefile and run.py files. I appreciate that having such makefile would be convenient to run some examples. Maybe we can have a separate PR to address it more properly .

@WeiluoRen-ByteDance Alright Wei Luo, I will remove the makefile and run.py in the top level. Regarding the logic in jaqmc/pp/ph/data.py, would you prefer that I open a separate PR to refactor it so that it matches the current data flow and address the other related issues?

File: jaqmc/loss/vmc.py (hunk 4 above)
Comment (Byte-RuichenLi) [reply jaqmc/loss/vmc.py:181]: In my experience, this shape-mismatch error typically occurs when the custom wavefunction, network(params, data), returns a (1,) tensor instead of a scalar. I suggest adding a check before the total_energy function, or simply updating line 69-70 from:

network = lambda *args, **kwargs: signed_network(*args, **kwargs)[1]
batch_local_energy = jax.vmap(sq_local_energy, in_axes=(None, 0, 0), out_axes=0)

to:

network = lambda *args, **kwargs: signed_network(*args, **kwargs)[1].squeeze()
scalar_local_energy = lambda *args, **kwargs: local_energy(*args, **kwargs).squeeze()
batch_local_energy = jax.vmap(scalar_local_energy, in_axes=(None, 0, 0), out_axes=0)

This would be a cleaner and more explicit fix.

**Comment by WeiluoRen-ByteDance** [comment]
@revpba Saw you just open a new PR (and thanks for that!). Closing the current one
